python/platforms/transition.bzl

```diff
@@ -0,0 +1,178 @@
+            is_executable = True,
+            output = zipfile_symlink,
+            target_file = zipfile,
+        )
+    env = {}
+    for k, v in ctx.attr.env.items():
+        env[k] = ctx.expand_location(v)
+    providers = [
+        DefaultInfo(
+            executable = executable,
+            files = depset([zipfile_symlink] if zipfile_symlink else [], transitive = [target[DefaultInfo].files]),
+            runfiles = ctx.runfiles([zipfile_symlink] if zipfile_symlink else []).merge(target[DefaultInfo].default_runfiles),
+        ),
+        target[PyInfo],
+        target[PyRuntimeInfo],
+        target[InstrumentedFilesInfo],
+        target[OutputGroupInfo],
+        # TODO(f0rmiga): testing.TestEnvironment is deprecated in favour of RunEnvironmentInfo but
+        # RunEnvironmentInfo is not exposed in Bazel < 5.3.
+        # https://github.com/bazelbuild/bazel/commit/dbdfa07e92f99497be9c14265611ad2920161483
+        testing.TestEnvironment(environment = env),
+    ]
+    return providers
+
+_COMMON_ATTRS = {
+    "env": attr.string_dict(
+        mandatory = False,
+    ),
+    "is_windows": attr.bool(
+        mandatory = True,
+    ),
+    "target": attr.label(
+        executable = True,
+        cfg = _transition_platform,
+        mandatory = True,
+        providers = [PyInfo],
+    ),
+    "target_platform": attr.label(
+        mandatory = True,
+    ),
+    # "tools" is a hack here. It should be "data" but "data" is not included by default in the
+    # location expansion in the same way it is in the native Python rules. The difference on how
+    # the Bazel deals with those special attributes differ on the LocationExpander, e.g.:
+    # https://github.com/bazelbuild/bazel/blob/ce611646/src/main/java/com/google/devtools/build/lib/analysis/LocationExpander.java#L415-L429
+    #
+    # Since the default LocationExpander used by ctx.expand_location is not the same as the native
+    # rules (it doesn't set "allowDataAttributeEntriesInLabel"), we use "tools" temporarily while a
+    # proper fix in Bazel happens.
+    #
+    # A fix for this was proposed in https://github.com/bazelbuild/bazel/pull/16381.
+    "tools": attr.label_list(
+        allow_files = True,
+        mandatory = False,
+    ),
+    # Required to Opt-in to the transitions feature.
+    "_allowlist_function_transition": attr.label(
+        default = "@bazel_tools//tools/allowlists/function_transition_allowlist",
+    ),
+}
+
+_transition_py_binary = rule(
+    _transition_py_impl,
+    attrs = _COMMON_ATTRS,
+    executable = True,
+)
+
+_transition_py_test = rule(
+    _transition_py_impl,
+    attrs = _COMMON_ATTRS,
+    test = True,
+)
+
+def _py_rule(rule, transition_rule, name, target_platform, **kwargs):
+    args = kwargs.pop("args", None)
+    data = kwargs.pop("data", None)
+    env = kwargs.pop("env", None)
+
+    # Attributes common to all build rules.
+    # https://bazel.build/reference/be/common-definitions#common-attributes
```

Comment:
Lets move the list of these attribute names into a separate file. It will most assuredly get reused by other code.